### DISCUSSION TITLE
App directory prefix for common files such as page.js or layout.js

### DISCUSSION THREAD

**Discussion by camsloanftc** [vercel/next.js#49645] [Ideas]
### Goals

Search-ability of common files in app directory.

### Background

The new app structure makes it harder to find things in your project when you have 20 files all named page.js, layout.js, etc.

### Proposal

It would be great if we could name/prefix these files, like:

- dashboard.page.js
- edit.page.ts
- auth.layout.ts
- blog-post.page.js, etc. 

and still have them picked up by the router system.

**Comment by icyJoseph** [comment]
You know, I was also somewhat concerned about this, but I learned to just search, for instance, `blog page`, and the two editors I use deal pretty well with it. See screenshots at the bottom.

That being said, I do like this, because it is able to built in meaning into the page name. I guess this has been discussed before though, there was a lot of fuzz when the `Layouts RFC` was released and the file names convention was introduced, and everyone threw in their folder structure knowledge/experiences/biases. Would be nice to hear from the Vercel team, now in 2023, with a stable router out there.

For prior art, we have the pages dir feature called Custom Page Extensions: https://nextjs.org/docs/pages/api-reference/next-config-js/pageExtensions

nvim:

![Screenshot 2023-05-11 at 08 05 47](https://github.com/vercel/next.js/assets/21013447/26a41953-0e06-43b4-9490-1749f37a81f0)

vscode:

![Screenshot 2023-05-11 at 08 06 09](https://github.com/vercel/next.js/assets/21013447/848a3f6c-0ccc-4f12-83e4-55df202fe39e)

**Comment by icyJoseph** [comment]
I'm also, somewhat thinking that this is in part a VSCode problem. One can configure the tab title label to short mode, which includes the parent directory name.

- https://stackoverflow.com/a/50181247

Why isn't this the default, beats me, but it sure helps a lot in several projects. Not only in Next.js.

**Comment by EternalWill43** [comment]
My eyes are over saturated with page.tsx, please strongly consider this proposal.

**Comment by webdevcody** [comment]
imagine browsing chrome and every tab says page, that's the next.js app router experience.

**Comment by mkorostoff** [comment]
This is a good idea! We [had this feature](https://nextjs.org/docs/pages/api-reference/next-config-js/pageExtensions) with the pages router before next13, and the documentation is not explicit that the feature was deliberately removed in the app router, so at very least I think this should be documented as a change in the latest release. I don't see any reason this feature couldn't co-exist with the app router though.

**Comment by mrksph** [comment]
Agree we need a way to differentiate at glance the file we are working on. Having just page.js is awful

**Comment by dariel87** [comment]
my first impression using next is page.tsx all over the tab in my vscode, hahaha. please consider this proposal

**Reply by icyJoseph** [reply]
You mean like, https://nextjs.org/docs/pages/api-reference/config/next-config-js/pageExtensions

That still works in App Route, but I guess it's not ideal, and I found an issue where people report some problems with it.

**Comment by inj-src** [comment]
![Screenshot 2025-06-11 023235](https://github.com/user-attachments/assets/89028102-1073-4b79-b1e3-3cc97c7b3343)
![Screenshot 2025-06-11 023253](https://github.com/user-attachments/assets/d10f2200-3511-4637-b40f-bfa2fc67c704)

**You can add this small snippet in the settings.json to achieve this.**

`   "workbench.editor.customLabels.patterns": {
      "**/app/**/{page,layout,route,error,loading}.{tsx,jsx}": "${dirname}/${filename}.${extname}",
      "**/app/**/[[]*[]]/{page,layout,route,error,loading}.{tsx,jsx}": "${dirname(1)}/${dirname}/${filename}.${extname}"
   },`
